### PR TITLE
feat(gatsby): Make __experimentalThemes part of plugin APIs

### PR DESCRIPTION
**Pull request by ChristopherBiscardi**
This change makes it so that what used to be called `__experimentalThemes`
is now part of the `plugins` APIs. This means that any plugin can take
advantage of the Theme APIs like Component Shadowing.

* This change contains no functional changes to how themes work today. The way that is accomplished is by "patching" plugins to look like the theme logic wants it to before passing it in. After we deprecate and remove `__experimentalThemes`, we can change this logic to be more plugin-specific.
* `__experimentalThemes` and `plugins` style themes can not be used at the
  same time. If `__experimentalThemes` is defined, it takes precedence
  so as to not break existing sites.

The plan here is to not break existing sites while enabling the stable
version of themes to be used. We (I most likely) should go back and remove
the `__experimentalThemes` handling after a suitable deprecation period.

**I'd like to see this released under a tag so we can test it on a wider range of sites before merging**

# TODO

- [ ] Add deprecation message for __experimentalThemes (should we do this in this PR or another?)
- [ ] Should we expand the webpack handling of themes in `packages/gatsby/src/utils/webpack.config.js` to work on all plugins? What's the status of https://github.com/gatsbyjs/gatsby/pull/14111 in relation to this?

### CODE REVIEW

**Comment by ChristopherBiscardi** [comment]
dupe fragments in the e2es

```
error GraphQL Error There was an error while compiling your site's GraphQL queries.
  Error: RelayParser: Encountered duplicate defintitions for one or more documents: each document must have a unique name. Duplicated documents:
- GatsbyImageSharpFixed
- GatsbyImageSharpFixed_tracedSVG
- GatsbyImageSharpFixed_withWebp
- GatsbyImageSharpFixed_withWebp_tracedSVG
- GatsbyImageSharpFixed_noBase64
- GatsbyImageSharpFixed_withWebp_noBase64
- GatsbyImageSharpFluid
- GatsbyImageSharpFluid_tracedSVG
- GatsbyImageSharpFluid_withWebp
- GatsbyImageSharpFluid_withWebp_tracedSVG
- GatsbyImageSharpFluid_noBase64
- GatsbyImageSharpFluid_withWebp_noBase64
- GatsbyImageSharpResolutions
- GatsbyImageSharpResolutions_tracedSVG
- GatsbyImageSharpResolutions_withWebp
- GatsbyImageSharpResolutions_withWebp_tracedSVG
- GatsbyImageSharpResolutions_noBase64
- GatsbyImageSharpResolutions_withWebp_noBase64
- GatsbyImageSharpSizes
- GatsbyImageSharpSizes_tracedSVG
- GatsbyImageSharpSizes_withWebp
- GatsbyImageSharpSizes_withWebp_tracedSVG
- GatsbyImageSharpSizes_noBase64
- GatsbyImageSharpSizes_withWebp_noBase64
```

**Comment by ChristopherBiscardi** [comment]
`e2e_tests_production_runtime` has a weird require error. not sure if this is an e2e-related setup issue or an actual themes issue

```
error Cannot find module 'gatsby-plugin-global-style'


  Error: Cannot find module 'gatsby-plugin-global-style'
  
  - loader.js:581 Function.Module._resolveFilename
    internal/modules/cjs/loader.js:581:15
  
  - v8-compile-cache.js:162 Function.require.resolve
    [production-runtime]/[v8-compile-cache]/v8-compile-cache.js:162:23
  
  - index.js:27 
    [production-runtime]/[gatsby]/dist/bootstrap/load-themes/index.js:27:43
  
  - Generator.next
  
  - new Promise
  
  - index.js:45 resolveTheme
    [production-runtime]/[gatsby]/dist/bootstrap/load-themes/index.js:45:17
  
  - index.js:111 
    [production-runtime]/[gatsby]/dist/bootstrap/load-themes/index.js:111:32
  
  - Generator.next
  
  - new Promise
```

**Comment by ChristopherBiscardi** [comment]
It looks like https://github.com/gatsbyjs/gatsby/pull/15179/files means we can remove the theme logic from the query compiler, which also fixes the duplicate fragments issue here

File: packages/gatsby/src/utils/webpack.config.js [hunk 1]
```diff
@@ -257,17 +257,7 @@ module.exports = async (program, directory, suppliedStage) => {
       rules.media(),
       rules.miscAssets(),
     ]
-    if (store.getState().themes.themes) {
```
Comment (wardpeet) [inline packages/gatsby/src/utils/webpack.config.js]: We need to keep this part of the plugin for now as I reverted full node_modules traversal from https://github.com/gatsbyjs/gatsby/pull/15270.

I'll remove this piece in a follow up PR where I traverse through plugins again.

**Comment by sidharthachatterjee** [comment]
https://github.com/gatsbyjs/gatsby/pull/15179 fixes the duplicated fragment issue

**Comment by sidharthachatterjee** [comment]
https://github.com/gatsbyjs/gatsby/pull/15284 should make a lot of the custom babel handling unnecessary

File: packages/gatsby/src/utils/webpack.config.js (hunk 1 above)
Comment (sidharthachatterjee) [reply packages/gatsby/src/utils/webpack.config.js]: Good to remove once https://github.com/gatsbyjs/gatsby/pull/15284 is in

**Comment by pieh** [comment]
Failing tests are result of duplicated plugin entries (fix in https://github.com/gatsbyjs/gatsby/pull/15307 )

**Comment by sidharthachatterjee** [comment]
Published in `gatsby@2.13.0` and holy crap, we have themes in stable now 💃

**Pull request merged**